src/librbd/librbd.cc

```diff
@@ -6140,36 +6140,55 @@ extern "C" int rbd_aio_write2(rbd_image_t image, uint64_t off, size_t len,
   return 0;
 }
 
-extern "C" int rbd_aio_writev(rbd_image_t image, const struct iovec *iov,
-                              int iovcnt, uint64_t off, rbd_completion_t c)
+static int get_iovec_length(const struct iovec *iov, int iovcnt, ssize_t &len)
 {
-  librbd::ImageCtx *ictx = (librbd::ImageCtx *)image;
-  librbd::RBD::AioCompletion *comp = (librbd::RBD::AioCompletion *)c;
-
-  // convert the scatter list into a bufferlist
-  auto aio_completion = get_aio_completion(comp);
-  ssize_t len = 0;
-  bufferlist bl;
+  len = 0;
   for (int i = 0; i < iovcnt; ++i) {
     const struct iovec &io = iov[i];
-    len += io.iov_len;
-    if (len < 0) {
-      break;
+    ssize_t len_tmp = len + io.iov_len;
+    if (len_tmp < len) {
+      return -EINVAL;
     }
+    len = len_tmp;
+  }
 
+  if (iovcnt <= 0) {
+    return -EINVAL;
+  }
+
+  return 0;
+}
+
+static bufferlist iovec_to_bufferlist(librbd::ImageCtx *ictx,
+                                      const struct iovec *iov,
+                                      int iovcnt,
+                                      librbd::io::AioCompletion* aio_completion)
+{
+  bufferlist bl;
+  for (int i = 0; i < iovcnt; ++i) {
+    const struct iovec &io = iov[i];
     bl.push_back(create_write_raw(ictx, static_cast<char*>(io.iov_base),
                                   io.iov_len, aio_completion));
   }
+  return bl;
+}
 
-  int r = 0;
-  if (iovcnt <= 0 || len < 0) {
-    r = -EINVAL;
-  }
+extern "C" int rbd_aio_writev(rbd_image_t image, const struct iovec *iov,
+                              int iovcnt, uint64_t off, rbd_completion_t c)
+{
+  librbd::ImageCtx *ictx = (librbd::ImageCtx *)image;
+  librbd::RBD::AioCompletion *comp = (librbd::RBD::AioCompletion *)c;
+
+  auto aio_completion = get_aio_completion(comp);
```

Comment:
Move this line inside the `if` body, immediately before the call to `iovec_to_bufferlist`.

Reply:
ack